### PR TITLE
security: Add seclink and HAL for a SE device

### PR DESCRIPTION
**Pull request by pillip8282**
1) Add seclink to communicate between kernel and user-space
2) Add security HAL to provide general abstraction layer to SE vendor

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [279dd09f18423d2c031e8374b0cf4de6e91c7e92] - Code Rule Check Result:
framework/src/seclink/seclink.c:41: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
framework/src/seclink/seclink.c:45: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
framework/src/seclink/seclink.c:46: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/drivers/seclink/seclink_drv_auth.c:52: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/drivers/seclink/seclink_drv_common.c:52: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/drivers/seclink/seclink_drv_common.c:52: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/drivers/seclink/seclink_drv_crypto.c:52: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/drivers/seclink/seclink_drv_ss.c:53: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/drivers/seclink/seclink_drv_utils.h:14: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/drivers/seclink/seclink_drv_utils.h:17: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:18: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:19: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:20: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:21: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:22: ERROR: [IDT_M_TAB] please, no space before tabs
os/drivers/seclink/seclink_drv_utils.h:23: ERROR: [IDT_M_TAB] please, no space before tabs

**Comment by seinfra** [comment]
Target : [279dd09f18423d2c031e8374b0cf4de6e91c7e92] - Code Rule Check (C++) OK.

File: os/se/Makefile [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+# -include $(TOPDIR)/.config
+-include $(TOPDIR)/Make.defs
+
+DEPPATH = --dep-path .
+SE_PATH		=
+SE_CSRCS	=
+SE_SRCS		=
+SE_LIB 	=
+SE_LIB_OBJS =
```
Comment (drashti304) [inline os/se/Makefile]: Please correct the indentations

Comment (pillip8282) [reply os/se/Makefile]: Thanks I'll fix it

**Comment by seinfra** [comment]
Target : [548bd62a106d3e5794815a0940f8d00c52ae384d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [548bd62a106d3e5794815a0940f8d00c52ae384d] - Code Rule Check (C++) OK.

File: os/drivers/seclink/seclink_drv.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		return -1;
+	}
+
+	/*  initialize upper */
+	char *drv_path = (char *)malloc(strlen(path) + 1);
+	if (!drv_path) {
+		free(upper);
+		return -1;
+	}
+	memcpy(drv_path, path, strlen(path) + 1);
```
Comment (chanijjani) [inline os/drivers/seclink/seclink_drv.c]: "drv_path" is used nowhere in the below. Is it correct..?

File: os/drivers/seclink/seclink_drv_key.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	if (!info) {
+		return -1;
+	}
+
+	struct sec_lowerhalf_s *se = (struct sec_lowerhalf_s *)lower;
+	if (!se || !(se->ops)) {
+		return -1;
+	}
+
+	printf("keymgr request cmd(%x) (%d)\n", cmd, info->mode);
```
Comment (chanijjani) [inline os/drivers/seclink/seclink_drv_key.c]: printf -> vdbg?

File: os/drivers/seclink/seclink_drv_ss.c [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	if (!info) {
+		return -1;
+	}
+
+	struct sec_lowerhalf_s *se = (struct sec_lowerhalf_s *)lower;
+	if (!se || !(se->ops)) {
+		return -1;
+	}
+
+	printf("[securestorage] request cmd(%x)\n", cmd);
```
Comment (chanijjani) [inline os/drivers/seclink/seclink_drv_ss.c]: printf -> vdbg?

File: os/drivers/seclink/seclink_drv.c (hunk 2 above)
Comment (pillip8282) [reply os/drivers/seclink/seclink_drv.c]: Thanks I'll fix

File: os/drivers/seclink/seclink_drv_ss.c (hunk 4 above)
Comment (pillip8282) [reply os/drivers/seclink/seclink_drv_ss.c]: Thanks I'll fix it

File: os/drivers/seclink/seclink_drv_key.c (hunk 3 above)
Comment (pillip8282) [reply os/drivers/seclink/seclink_drv_key.c]: Thanks I'll fix it

**Comment by seinfra** [comment]
Target : [657f323674baedd7f065cc411e0dfcadc8f50737] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [657f323674baedd7f065cc411e0dfcadc8f50737] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [bcd159ad12b194f3e38b3408c6b46ea6a92c817f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bcd159ad12b194f3e38b3408c6b46ea6a92c817f] - Code Rule Check OK.

**Pull request closed without merging**